lapack/gonum/dgghrd.go

```diff
@@ -0,0 +1,124 @@
+// Copyright ©2023 The Gonum Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+
+package gonum
+
+import (
+	"gonum.org/v1/gonum/blas"
+	"gonum.org/v1/gonum/blas/blas64"
+	"gonum.org/v1/gonum/lapack"
+)
+
+// Dgghrd reduces a pair of real matrices (A,B) to generalized upper
+// Hessenberg form using orthogonal transformations, where A is a
+// general matrix and B is upper triangular.  The form of the
+// generalized eigenvalue problem is
+//
+//	A*x = lambda*B*x,
+//
+// and B is typically made upper triangular by computing its QR
+// factorization and moving the orthogonal matrix Q to the left side
+// of the equation.
+// This subroutine simultaneously reduces A to a Hessenberg matrix H:
+//
+//	Qᵀ*A*Z = H
+//
+// and transforms B to another upper triangular matrix T:
+//
+//	Qᵀ*B*Z = T
+//
+// in order to reduce the problem to its standard form
+//
+//	H*y = lambda*T*y
+//
+// where y = Zᵀ*x.
+//
+// The orthogonal matrices Q and Z are determined as products of Givens
+// rotations.  They may either be formed explicitly, or they may be
+// postmultiplied into input matrices Q1 and Z1, so that
+//
+//	Q1 * A * Z1ᵀ = (Q1*Q) * H * (Z1*Z)ᵀ
+//	Q1 * B * Z1ᵀ = (Q1*Q) * T * (Z1*Z)ᵀ
+//
+// If Q1 is the orthogonal matrix from the QR factorization of B in the
+// original equation A*x = lambda*B*x, then Dgghrd reduces the original
+// problem to generalized Hessenberg form.
+//
+// Dgghrd is an internal routine. It is exported for testing purposes.
+func (impl Implementation) Dgghrd(compq, compz lapack.OrthoComp, n, ilo, ihi int, a []float64, lda int, b []float64, ldb int, q []float64, ldq int, z []float64, ldz int) {
+	switch {
+	case (compq != lapack.OrthoNone && compq != lapack.OrthoEntry && compq != lapack.OrthoUnit) ||
+		(compz != lapack.OrthoNone && compz != lapack.OrthoEntry && compz != lapack.OrthoUnit):
+		panic(badOrthoComp)
+	case len(a) < (n-1)*lda+n:
+		panic(shortA)
+	case len(b) < (n-1)*ldb+n:
+		panic(shortB)
+	case n < 0:
+		panic(nLT0)
+	case ilo < 0:
+		panic(badIlo)
+	case ihi < ilo-1 || ihi >= n:
+		panic(badIhi)
+	case lda < max(1, n):
+		panic(badLdA)
+	case ldb < max(1, n):
+		panic(badLdB)
+	case (compq != lapack.OrthoNone && ldq < n) || ldq < 1:
+		panic(badLdQ)
+	case (compz != lapack.OrthoNone && ldz < n) || ldz < 1:
+		panic(badLdZ)
+	case compq != lapack.OrthoNone && len(q) < (n-1)*ldq+n:
+		panic(shortQ)
+	case compz != lapack.OrthoNone && len(z) < (n-1)*ldz+n:
+		panic(shortZ)
+	}
+
+	if compq == lapack.OrthoUnit {
+		impl.Dlaset(blas.All, n, n, 0, 1, q, ldq)
+	}
+	if compz == lapack.OrthoUnit {
+		impl.Dlaset(blas.All, n, n, 0, 1, z, ldz)
+	}
+	if n == 0 {
+		return // Quick return if possible.
+	}
+
+	// Zero out lower triangle of B.
+	for i := 1; i < n; i++ {
+		for j := 0; j < i-1; j++ {
```

Comment:
```suggestion
		for j := 0; j < i; j++ {
```
Please modify the test to catch this.

Reply:
Done!

Reply:
I'm not so happy about the new testing helper and about using NaN so much in the tests. Since it's mostly nitpicking, I'll merge this and then submit again a cleanup PR later.